ocaml/typing/typecore.ml

```diff
@@ -3342,6 +3348,12 @@ let collect_unknown_apply_args env funct ty_fun mode_fun rev_args sargs ret_tvar
 let collect_apply_args env funct ignore_labels ty_fun ty_fun0 mode_fun sargs ret_tvar =
   let warned = ref false in
   let rec loop ty_fun ty_fun0 mode_fun rev_args sargs =
+    let type_unknown_args () =
+      (* We're not looking at a *known* function type anymore, or there are no
+         arguments left. *)
+      collect_unknown_apply_args env funct ty_fun0 mode_fun rev_args sargs ret_tvar
+    in
+    if sargs = [] then type_unknown_args () else
```

Comment:
I think this line is not needed, because in the immediately following match, `sargs` is matched as well. In that case, `type_unknown_args` is used only once - maybe we don't need a function for it?

Reply:
The idea is to skip the `expand_head` if the list is empty, which is extra work and changes which programs are accepted. I did this only because the upstream patch does the same; this is actually responsible for the diff in `ocaml/testsuite/tests/typing-gadts/ambivalent_apply.ml` that also shows up upstream.

Reply:
That makes sense - thanks.